### DISCUSSION TITLE
Where can I get the list of available validators?

### DISCUSSION THREAD

**Discussion by karthurdnt** [beeware/toga#1451] [Q&A] [answered]
If someone could help me find the list of textinput validators, I'd be grateful. 
I know they exist they are mentioned in the documentation (https://toga.readthedocs.io/en/latest/reference/api/widgets/textinput.html) and in some issues here I've read they were implemented with some default validators, but I couldn't find the list of all of them.

**Comment by freakboy3742** [comment, marked as answer]
Unfortunately, they're not independently documented at present; but they're all defined in [the validators module](https://github.com/beeware/toga/blob/master/src/core/toga/validators.py) of toga-core - so they can be imported as `from toga.validators import <whatever>`.

**Reply by karthurdnt** [reply]
Thank you, that's what I needed! I am trying them now...

One more quick question: how do I extract those error_messages values so I can put it into a warning window? I tried the .validate() method but I just get the Booleans from that.

**Reply by freakboy3742** [reply]
Take a look at the [textinput example (examples/textinput in the toga repository) for a demonstration of how to use validators in practice.

**Reply by karthurdnt** [reply]
I took a look there, but I was not getting the error messages, the terminal would output:
 **_[GTK+] Not implemented: TextInput.set_error()_**

So I've made a tiny little change to the source code to be able to extract the error_message, changing it to an instance variable. I'm not sure if it's worth a PR, but it solved what I wanted.

**Reply by freakboy3742** [reply]
PRs are always welcome - no
matter how small. A tiny improvement is still an improvement!

**Reply by aendie** [reply]
the link to "the validators module" of toga core (above) returns "404 - page not found"

**Reply by freakboy3742** [reply]
The file that is referenced has been moved to [here](https://github.com/beeware/toga/blob/main/core/src/toga/validators.py); GitHub isn't resolving the change of location. 

However, since that was posted, we've also [added documentation](https://toga.beeware.org/en/stable/reference/api/resources/validators.html).
